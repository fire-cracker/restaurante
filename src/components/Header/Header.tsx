import React from 'react'
import Nav from 'react-bootstrap/Nav'
import Navbar from 'react-bootstrap/Navbar'
import Button from 'react-bootstrap/Button'

const Header = () => {
  return (
    <Navbar bg='black' expand='lg' variant="dark" className='sticky-top'>
      <Navbar.Brand href='/home'>React-Bootstrap</Navbar.Brand>
      <Navbar.Toggle aria-controls='basic-navbar-nav' />
      <Navbar.Collapse id='basic-navbar-nav'>
        <Nav activeKey='/' className='m-auto main-tabs'>
          <Nav.Link href='/'>HOME</Nav.Link>
          <Nav.Link href='/menu'>MENU</Nav.Link>
          <Nav.Link href='/menu'>BOOK A TABLE</Nav.Link>
        </Nav>
        <Nav className='mr'>
          <Button href="/login" variant="outline-darkkhaki">LOGIN</Button>
          <Button href='/profile' variant="outline-transparent"><i className="fa fa-shopping-cart text-white"></i></Button>
        </Nav>
      </Navbar.Collapse>
    </Navbar>
  )
}

export default Header
