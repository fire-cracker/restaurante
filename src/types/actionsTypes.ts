export interface Action<T> {
  type: ActionType;
  payload?: T;
}

export enum ActionType {
  SIGNUP_REQUEST_PENDING = 'SIGNUP_REQUEST_PENDING',
  SIGNUP_REQUEST_SUCCESS = 'SIGNUP_REQUEST_SUCCESS',
  SIGNUP_REQUEST_FAILED = 'SIGNUP_REQUEST_FAILED',
  
  LOGIN_REQUEST_PENDING = 'LOGIN_REQUEST_PENDING',
  LOGIN_REQUEST_SUCCESS = 'LOGIN_REQUEST_SUCCESS',
  LOGIN_REQUEST_FAILED = 'LOGIN_REQUEST_FAILED',

  GET_PROFILE_REQUEST_PENDING = 'GET_PROFILE_REQUEST_PENDING',
  GET_PROFILE_REQUEST_SUCCESS = 'GET_PROFILE_REQUEST_SUCCESS',
  GET_PROFILE_REQUEST_FAILED = 'GET_PROFILE_REQUEST_FAILED',

  LOGOUT_REQUEST_SUCCESS = 'LOGOUT_REQUEST_SUCCESS',

  SET_USER_STATE = 'SET_USER_STATE'
}