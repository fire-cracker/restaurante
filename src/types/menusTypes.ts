export interface IMenu {
  id: number
  name: string
  image: string
  price: number
  recipe: string
  type: string
}

export interface IMenuState {
  menus: IMenu | null
  count: number
  fetching: boolean
  fetched: boolean
}
