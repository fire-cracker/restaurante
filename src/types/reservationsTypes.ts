export interface IReservation {
  id: number
  userId: string
  date: string
  time: string
  type: string
  price: number
  persons: number
}
export interface INewReservation {
  date: string
  time: string
  type: string
  persons: number
}

export interface IReservationState {
  reservation: IReservation | null
  fetching: boolean
  fetched: boolean
}

export interface IStripeCharge {
  id: string
  object: string
  amount: 2000
  paid: boolean
  receipt_email: string
  receipt_number: string
  receipt_url: string
  status: string
}
