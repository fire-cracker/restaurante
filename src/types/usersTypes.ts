export interface IUser {
  id: string
  username: string
  email: string
  role: 'customer' | 'admin'
}

export interface IUserState {
  user: IUser | null
  logingIn: boolean
  isLoggedIn: boolean
}