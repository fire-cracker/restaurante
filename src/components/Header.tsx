import React, { FC, ReactElement } from 'react'
import { withRouter, Link } from 'react-router-dom'
import { RouteComponentProps } from 'react-router'
import { connect } from 'react-redux'
import Nav from 'react-bootstrap/Nav'
import Navbar from 'react-bootstrap/Navbar'
import Button from 'react-bootstrap/Button'

import { IRootState } from '../redux/reducers'
import { IUserState } from '../types/usersTypes'
import { logout } from '../redux/actions/users'

interface Props extends RouteComponentProps<void> {
  executeScrollToMenu: (history: any) => void
  history: any
  onModalShow: () => void
  userState: IUserState
  logout: () => void
}
const Header: FC<Props> = ({
  history,
  executeScrollToMenu,
  onModalShow,
  userState,
  logout
}): ReactElement => (
  <Navbar collapseOnSelect bg="black" expand="lg" variant="dark" id="nav" className="sticky-top">
    <Navbar.Brand href="/">Restaurant√©</Navbar.Brand>
    <Navbar.Toggle aria-controls="basic-navbar-nav" />
    <Navbar.Collapse id="basic-navbar-nav">
      <Nav defaultActiveKey="home" className="m-auto main-tabs">
        <Nav.Link as={Link} to="/" eventKey="home">
          HOME
        </Nav.Link>
        <Nav.Link as={Link} to="/menu" eventKey="menu" onClick={() => executeScrollToMenu(history)}>
          MENU
        </Nav.Link>
        <Nav.Link as={Link} eventKey="reservation" to="/reservation">
          BOOK A TABLE
        </Nav.Link>
      </Nav>
      <Nav className="mr">
        {userState.isLoggedIn ? (
          <Button href="#" variant="outline-darkkhaki" onClick={() => logout()}>
            LOGOUT
          </Button>
        ) : (
          <Button href="#" variant="outline-darkkhaki" onClick={() => onModalShow()}>
            LOGIN
          </Button>
        )}
      </Nav>
    </Navbar.Collapse>
  </Navbar>
)

const mapStateToProps = (state: IRootState) => ({ userState: state.userState })
const mapDispatchToProps = { logout }

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Header))
