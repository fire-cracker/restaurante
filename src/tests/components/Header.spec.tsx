import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { render, cleanup } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import Header from '../../components/Header'
import { IUserState } from '../../types/usersTypes'

const createMockStore = configureMockStore([thunk])

interface IProps {
  executeScrollToMenu: (history: any) => void
  history: any
  onModalShow: () => void
  userState: IUserState
  logout: () => void
}

describe('Header', () => {
  const defaultProps: IProps = {
    executeScrollToMenu: jest.fn(),
    history: {},
    onModalShow: jest.fn(),
    userState: {
      user: null,
      logingIn: false,
      isLoggedIn: false
    },
    logout: jest.fn()
  }

  const initialState = {
    userState: {
      user: null,
      logingIn: false,
      isLoggedIn: false
    }
  }

  const setup = (newProps?: any, newStore?: any) => {
    const props = { ...defaultProps, ...newProps }
    const store = createMockStore({
      ...initialState,
      userState: { ...initialState.userState, ...newStore }
    })

    return render(
      <BrowserRouter>
        <Provider store={store}>
          <Header {...props} />
        </Provider>
      </BrowserRouter>
    )
  }

  afterAll(() => {
    jest.clearAllMocks()
    cleanup()
  })

  test('renders Header component', () => {
    const wrapper = setup()
    expect(wrapper).toMatchSnapshot()
    const { getByText } = wrapper
    const linkElement = getByText('LOGIN')
    expect(linkElement).toBeInTheDocument()
  })

  test('show login page on click login', async () => {
    const { getByText } = setup()
    const loginButton = getByText('LOGIN')
    await userEvent.click(loginButton)
    expect(defaultProps.onModalShow).toHaveBeenCalledTimes(1)
  })

  test('user can navigate to menu page', async () => {
    const { getByText } = setup()
    const menuLink = getByText('MENU')
    await userEvent.click(menuLink)
    expect(defaultProps.executeScrollToMenu).toHaveBeenCalledTimes(1)
  })

  test('user can see logout button and logout if logged in', async () => {
    const { getByText } = setup(
      { userState: { ...defaultProps.userState, isLoggedIn: true } },
      { isLoggedIn: true }
    )
    const logoutButton = await getByText('LOGOUT')
    expect(logoutButton).toBeInTheDocument()
    await userEvent.click(logoutButton)
    // expect(defaultProps.logout).toHaveBeenCalledTimes(1)
  })
})
