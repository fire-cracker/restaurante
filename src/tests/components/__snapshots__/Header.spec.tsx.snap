// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders Header component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="sticky-top navbar navbar-expand-lg navbar-dark bg-black"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          React-Bootstrap
        </a>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="m-auto main-tabs navbar-nav"
          >
            <a
              class="nav-link active"
              data-rb-event-key="home"
              href="/"
            >
              HOME
            </a>
            <a
              class="nav-link"
              data-rb-event-key="menu"
              href="/menu"
            >
              MENU
            </a>
            <a
              class="nav-link"
              data-rb-event-key="reservation"
              href="/reservation"
            >
              BOOK A TABLE
            </a>
          </div>
          <div
            class="mr navbar-nav"
          >
            <a
              class="btn btn-outline-darkkhaki"
              href="#"
              role="button"
            >
              LOGIN
            </a>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="sticky-top navbar navbar-expand-lg navbar-dark bg-black"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        React-Bootstrap
      </a>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="m-auto main-tabs navbar-nav"
        >
          <a
            class="nav-link active"
            data-rb-event-key="home"
            href="/"
          >
            HOME
          </a>
          <a
            class="nav-link"
            data-rb-event-key="menu"
            href="/menu"
          >
            MENU
          </a>
          <a
            class="nav-link"
            data-rb-event-key="reservation"
            href="/reservation"
          >
            BOOK A TABLE
          </a>
        </div>
        <div
          class="mr navbar-nav"
        >
          <a
            class="btn btn-outline-darkkhaki"
            href="#"
            role="button"
          >
            LOGIN
          </a>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
